kernel: add a config switch for process debug info

alphan opened the pull request:
### Pull Request Overview

This change adds a config switch to control the collection of debugging information for processes and improve syscall latency.

I also added the `debug_` prefix to two functions and refactored some code to new `debug_*` functions.

See also: #1730, #1443.

### Testing Strategy

I ran the app used to test #1826 with `debug_processes` both `true` and `false`. Setting `debug_processes = false` saves ~60 cycles when `opt-level=3`, `lto=true`, `debug=false` and ~200 cycles when `codegen-units=1` in addition to the previous settings.

### TODO or Help Wanted

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make formatall`.

hudson-ayers commented on kernel/src/process.rs:
```diff
@@ -1804,8 +1847,9 @@ impl<C: 'static + Chip> Process<'a, C> {
                     .current_stack_pointer
                     .set(new_stack_pointer as *mut u8);
-                process.debug_set_max_stack_depth();
-                process.stored_state.set(stored_state);
```
I think this line needs to stay (though this will be moot when your other PR is merged).

alphan replied:
Yes, maybe I should have included the commit in #1826 in this PR too.

alphan replied:
Done.

bradjc commented:
> I also think that when this config option is enabled, the process console should report that the values returned are inaccurate, and can be enabled by toggling this option in the kernel.

This is one example of why these config options get scary...reasoning about what works and doesn't work with every single combination of config options, and trying to correctly handle every case is very difficult to scale.

alphan commented:
- `debug_processes` is now on by default.
- Added warning messages in process console (list command output) and `print_full_process()`.
  - Introduced an associated function (could be an associated constant too, not sure which one is preferred) to make `debug_processes` available in the process console.

alphan commented:
> > I also think that when this config option is enabled, the process console should report that the values returned are inaccurate, and can be enabled by toggling this option in the kernel.
> 
> This is one example of why these config options get scary...reasoning about what works and doesn't work with every single combination of config options, and trying to correctly handle every case is very difficult to scale.

I agree that these config options can have unintended consequences, especially when they break the assumptions of existing clients as in this case, but hopefully the latest changes address @hudson-ayers's comments.

gendx commented on kernel/src/process.rs:
```diff
@@ -1528,6 +1570,15 @@ impl<C: Chip> ProcessType for Process<'a, C> {
     }
 
     unsafe fn print_full_process(&self, writer: &mut dyn Write) {
+        if config::CONFIG.debug_processes == false {
+            let _ = writer.write_str(
+                "\
+                \r\nWARNING: Values reported below are inaccurate because\
+                \r\nprocess debugging is currently disabled. This setting\
+                \r\ncan be enabled in `kernel/src/config.rs`.\
+                \r\n",
+            );
+        }
```
If `debug_processes` is disabled, could this function simply write this banner and skip all of the rest? This would also reduce code size because the rest of the function would be dead code.

Also, will this banner be printed once per process, or once for all processes?

alphan replied:
> If debug_processes is disabled, could this function simply write this banner and skip all of the rest? This would also reduce code size because the rest of the function would be dead code.

It could, but this function also prints some other useful information that is not affected by `debug_processes`, e.g. stored state of the process and MPU config.


> Also, will this banner be printed once per process, or once for all processes?

This banner will be printed per process. I can move the banner to `panic_process_info` since it is the only caller for now but I thought keeping it here would be safer, just in case this function is called from somewhere else in the future. It is also harder to miss when there is a lot of debug output.

Please let me know what you think.

gendx replied:
I thought that printing the banner once would be enough, and avoid verbosity, but if you think it's better to print it multiple times to avoid missing it I don't have a strong opinion either.

bradjc commented on kernel/src/process.rs:
```diff
@@ -431,6 +441,13 @@ pub trait ProcessType {
+
+    /// Collect information about context switches. Currently only counts
+    /// the number of context switches due to timeslice expirations.
+    fn debug_context_switch(&self, switch_reason: &Option<syscall::ContextSwitchReason>);
```
Many of these do not need to be in the trait. If they aren't called from outside process.rs we can just implement them on the struct.

alphan replied:
Done. Thanks for pointing this out, this lead me to a rabbit hole of compiler settings :) Please see my comment above.

bradjc commented on kernel/src/sched.rs:
```diff
@@ -284,6 +284,14 @@ impl Kernel {
+    ///
+    /// Clients that rely on this information, e.g. process console, should call this method
+    /// to check if process debugging is enabled and warn the user if it is not.
+    pub fn process_debugging_enabled() -> bool {
```
Rather than this, which is very easy to forget to call, what about making the debug functions that can return useless values return `None`?

alphan replied:
I added this to be able to print the warning banner in the process console.

`debug_*` functions currently return `0` if `Process::debug` is not there when `debug_processes` is disabled.  It looks like returning `None` instead of `0` would result in a lot of changes because callers expect `usize` from these functions and it could still possibly be misinterpreted by someone not familiar with the output. Please let me know what you think.

gendx replied:
If the kernel config needs to be exposed outside of the kernel crate, shouldn't we make the config module public, or expose a function returning a (read-only) reference to `config::CONFIG`?

hudson-ayers replied:
Is it safe to assume that anything in config::CONFIG should be public to capsules? 

Though I agree that a function returning a specific configuration feature probably belongs in `config.rs`, not `sched.rs`. I suppose that an alternative would be to make `Config` a pub struct, but then internally use the `crate` modifier on any fields we don't want exposed to capsules?

alphan replied:
@gendx, I agree but I just wasn't sure how much of it should be exposed so I exposed the new flag only. The function is implemented in `kernel/src/sched.rs` because I thought `Kernel` was a reasonable home for a kernel config but if it is OK to make `Config` a pub struct, I can definitely implement @hudson-ayers's suggestion.

gendx replied:
I think @hudson-ayers's suggestion would be the way to go. By the way, we should use `pub(crate)` now, as was finally decided in https://github.com/tock/tock/pull/1650#issuecomment-629399631.

On the other hand, it seems to me that instead of exposing the config to capsules, we could just add a function to the kernel that contains the following code, and only expose this function for capsules.

https://github.com/tock/tock/blob/44875ac55169c99bcb15222e2c84d0726dce3d2f/capsules/src/process_console.rs#L252-L255

The code and text for this banner is already duplicated anyway, so having a single function would also reduce the size of the binary.

https://github.com/tock/tock/blob/44875ac55169c99bcb15222e2c84d0726dce3d2f/kernel/src/process.rs#L1412-L1420

bradjc replied:
I do think it would require at least some changes, but if a `debug_` call can now fail when before it couldn't, I think that should be reflected in its return type. Expecting callers to first check a flag in a completely different file before knowing if a return value is valid seems like an API we don't want.

brghena commented on kernel/src/sched.rs:
```diff
@@ -364,7 +374,9 @@ impl Kernel {
-                            process.debug_syscall_called(syscall);
+                            if config::CONFIG.debug_processes {
+                                process.debug_syscall_called(syscall);
+                            }
```
If so, it might be more straightforward to relocate all of the `if config...` checks into the debug functions themselves rather than having them sprinkled throughout the code.

alphan replied:
I actually tried this before creating this PR but tried again after implementing @bradjc's comment below to make sure. Benchmarking something like this is tricky but IIUC it doesn't work for functions that are on the trait but it works when they are implemented on the struct. 
This conclusion is not based on the number of cycles, because it depends on compiler settings and can be misleading, but `objdump` output. I see calls to `debug_*` functions that are on the trait when `debug_processes = false` (even when they have an early return) unless they are wrapped in an if statement. Thus, I was able to move the if statements into the `debug_*` functions implemented on the struct but kept the rest (for calls outside process.rs). I hope this is an acceptable compromise.

gendx replied:
I don't think functions on the `ProcessType` trait can be effectively optimized away by the compiler, because this trait is currently used as trait objects, that have dynamic dispatch (even though in practice there's only one type per compiled board).

If you're interested, I've started documenting this overhead in #1761 (see also #1663).
Removing `&dyn` on some GPIOs types recently led to code size improvements in #1765.

alphan replied:
I find it natural that the functions on the trait remain, but I am not sure why the calls are not optimized away. Maybe they are optimized in other cases as a side-effect? Thanks for the issue numbers, I'll take a look!

alphan commented:
- Moved some `debug_*` functions from the trait to the struct, added early returns, and removed if statements from call sites.
- Rebased on master.

brghena commented on kernel/src/process.rs:
```diff
@@ -1986,4 +1966,64 @@ impl<C: 'static + Chip> Process<'a, C> {
     }
+
+    /// Update the maximum stack depth that the process has ever used.
+    fn debug_update_max_stack_depth(&self) {
+        if config::CONFIG.debug_processes == false {
+            return;
+        }
+        self.debug_update_max_stack_depth_using_pointer(self.current_stack_pointer.get());
+    }
```
Does this compile differently than:

```rust
    fn debug_update_max_stack_depth(&self) {
        if config::CONFIG.debug_processes {
             self.debug_update_max_stack_depth_using_pointer(self.current_stack_pointer.get());
        }
    }
```

Generally, I'd say writing it this way reads clearer.

alphan replied:
Done.

alphan commented:
Implemented @hudson-ayers's and @gendx's comments (removed `process_debugging_enabled()` and exposed `debug_processes` to capsules).

@gendx, I think capsules may want to do different things based on the value of this flag instead of just printing a warning. So, I thought it would be better to expose this flag. Re. duplication, how about moving the message to a function that returns a static str? Also, where should such a function be defined?

gendx commented:
I think the main concern about exposing the configuration outside of the kernel would be that it could be harder to refactor in the future if capsules start relying on it. So I see the argument of keeping it private unless putting it in the public API is really necessary.

phil-levis commented:
Debugging information is probably one of the toughest of these (config option questions), because you want to be able to selectively choose outputs depending on what you are debugging. Code space means you can't just always leave it on.

The process console is a capsule, though. We find it useful, but it shouldn't be that everyone has to pay a tax for it. None seems like avoiding the problem: we should be able to detect this kind of mismatch at compile time. Else we are paying a big cost for this super-long string that's explaining our configuration mismatch. Is there some way to redesign this interface so we can detect it at compile-time (e.g., the absence of a trait or module)?

bradjc commented:
Moved to #1995
